### ISSUE TITLE
Profiling with MiniEngine

### ISSUE THREAD

**Issue by jpvanoosten** [Microsoft/DirectX-Graphics-Samples#59]
I noticed there are profiling features built-in to MiniEngine. I'm fairly new to profiling graphics applications in Visual Studio so I was wondering if there is any documentation or examples showing how to use the profiling tools that are made available in MiniEngine?

I was looking for some documentation in the [GitHub Wiki for MiniEngine](https://github.com/Microsoft/DirectX-Graphics-Samples/wiki) but there is no documentation there.

Specifically I would like to perform a comparative performance analysis on the various stages of rendering. For example, determine the average time to perform a specific post-process effects over several seconds of rendering. There doesn't seem to be an easy way to determine this information using the GPU profiling tools in Visual Studio (or I have not found an easy way to do this).

**Comment by bobbrow** [comment]
Do you mean that you saw the profiling features in the code, or that you used the game controls to pop up the debug overlay?  If you are looking for profiling information about the various stages of rendering, you can enable the "Display Profiler" option in the debug overlay and you will get a grid of information that you can navigate with the arrow keys once the debug overlay is closed.

* to open the debug overlay: back button (game pad) or backspace (keyboard)
* to navigate the debug menu: dpad (game pad) or arrow keys (keyboard)
* to toggle a debug menu item: A button (game pad) or return (keyboard)

Are you looking to perform comparative analysis of your own app, or of MiniEngine?  In either case you should look at the GpuTimeManager class.  This illustrates how to use time stamp queries.  You basically add a time stamp query into the command list before adding your rendering stage's commands, then you add one more query right after those commands.  When you are ready to process the results, you add a Resolve command to the command list and read back the results by Mapping the READBACK buffer storing your query results and finding the difference between the two time stamps.

And one last thing, Visual Studio 2015 Update 1 should be out at the end of the month which includes improved GPU Profiling tools.  You will also want to have the fall update for Windows 10 (build 10586) installed too and ensure the latest Graphics Tools optional feature is installed.  After you update, you can use ALT+F2 to run your app and capture all of the timing data about your command lists (individual draw calls, etc) and should make it easier to visualize your GPU overhead without needing to add all of the time stamp query code into your app.  However, it won't do averaging from what I can tell, but you can export results to a spreadsheet for further analysis.

**Comment by jpvanoosten** [comment]
Bob,

I'm using MiniEngine for a graphics experiment that I am doing for my master thesis. Eventually, I will want to be able to record the timings of various stages of my implementation. I'm keeping a blog here: [3dgep.blogspot.com](http://3dgep.blogspot.nl/) in case you are curious about my progress with MiniEngine.

I noticed that MiniEngine had support for adding profiling markers using the PIX library but it wasn't immediately obvious how to display the CPU/GPU timings. I was trying to use the built-in VS2015 (with Update 1 RC) profiler  and I noticed the nice markers in the event list but getting average timings proved to be slightly more difficult. I suppose I could use the Excel export but since I would have to take multiple timings this could become cumbersome.

Thanks for your tip about the on-screen timings. That was what I was looking for but I totally missed the fact that after I enabled the "Display Profiler" option, I had to fist close the debug overlay before the profile data appeared!
